Detox 20 (#3285)
wix/Detox: docs/introduction/project-setup.mdx

import CodeBlock from '@theme/CodeBlock';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

import PrepareAppConfigsIos from './partials/_project-setup-apps-ios.mdx';
import PrepareAppConfigsAndroid from './partials/_project-setup-apps-android.mdx';
import PrepareDeviceConfigsIos from './partials/_project-setup-devices-ios.mdx';
import PrepareDeviceConfigsAndroid from './partials/_project-setup-devices-android.mdx';

# Project Setup

:::info

This article mainly covers standard React Native projects.

> If some steps don't look applicable to your project, please adapt them accordingly, relying on the common sense –

Reviewer: ```suggestion
If some steps don't look applicable to your project, please adapt them accordingly, relying on common sense –
```

  </TabItem>
</Tabs>

```bash npm2yarn
npm install detox@20.0.12-prerelease.0 --save-dev
```

The next step would be to **install a test runner**. Out of the box, Detox offers you first-class integration
with [Jest](https://jestjs.io), and that is our official recommendation:

```bash npm2yarn
npm install "jest@>=27.2.5" --save-dev
```

:::info
The command above will install the **latest Jest version** regardless of the constraint, but `@>=27.2.5` is
recommended in the first place to prevent you from getting an incompatible Jest version unknowingly, under influence
of the existing lock files. Reportedly, `package-lock.json` could restrict `npm install jest` to installing
an older, unsupported version `26.x`.
:::

Now, when all the dependencies are installed, initialize Detox in your project:

```bash
detox init
```

Normally you should see an output like:

```plain text
Created a file at path: .detoxrc.js
Created a file at path: e2e/jest.config.js
Created a file at path: e2e/starter.test.js
```

:::tip
If you see a message like `command not found: detox`, make sure you have [installed Detox command line tools](#1-command-line-tools-detox-cli).
:::

> After Detox generated these files in your project's root, you still have some work to do with them:

Reviewer: ```suggestion
After Detox generates these files in your project's root, you still have some work to do with them:
```

* **Manifests:**
    * `android/app/src/main/AndroidManifest.xml`
    * `android/app/src/main/res/xml/network_security_config.xml`

### 3.1. Patching build scripts

Open your main build script and apply the suggested changes:

```diff title="android/build.gradle"
 buildscript {
   ext {
     buildToolsVersion = "31.0.0"
// highlight-next-line
     minSdkVersion = 21 // (1)
     compileSdkVersion = 30
     targetSdkVersion = 30
// highlight-next-line
+    kotlinVersion = 'X.Y.Z' // (2)
   }
 …
   dependencies {
     classpath("com.android.tools.build:gradle:7.1.1")
     classpath("com.facebook.react:react-native-gradle-plugin")
     classpath("de.undercouch:gradle-download-task:5.0.1")
// highlight-next-line
+    classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlinVersion") // (3)
 …

 allprojects {
   repositories {
   …
   google()
// highlight-start
+  maven { // (4)
+    url("$rootDir/../node_modules/detox/Detox-android")
+  }
// highlight-end
   maven { url 'https://www.jitpack.io' }
 }
```

Here is the explanation of those changes, numbered 1-4:

1. Make sure your `minSdkVersion` is at least 18 or higher.

> 1. We recommend you to define a global `kotlinVersion` constant you would use in the next step.

Reviewer: ```suggestion
1. We recommend that you define a global `kotlinVersion` constant that you can use in the next step.
```

Replace `X.Y.Z` with the actual version number.
To get one, open **Android Studio**, go to `Preferences > Languages & Frameworks > Kotlin` and
look at `Current Kotlin plugin version` field. For example, `211-1.5.30-release-408-AS7442.40`
means you have version `1.5.30`.
    ![](../img/android/get-kotlin-version.png)
1. The line adds [Kotlin Gradle plugin](https://mvnrepository.com/artifact/org.jetbrains.kotlin/kotlin-gradle-plugin)
to the build script. If your project is not entirely new, there's a chance you might have it already, so make sure
you aren't adding it twice. :wink:

> 1. Last, we are adding Detox as a precompiled native dependency, `.aar`. If you need to have Detox as a

Reviewer: ```suggestion
1. Last, we add Detox as a precompiled native dependency, `.aar`. If you need to have Detox as a
```

Now let's move on to the next build script and prepare it:

```diff title="android/app/build.gradle"
 …

 android {
   …
   defaultConfig {
     …
     versionCode 1
     versionName "1.0"
+    testBuildType System.getProperty('testBuildType', 'debug')
+    testInstrumentationRunner 'androidx.test.runner.AndroidJUnitRunner'
   …
   buildTypes {
     release {
       minifyEnabled true
       proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+      proguardFile "${project(':detox').projectDir}/proguard-rules-app.pro"

       signingConfig signingConfigs.release
     }
   }
   …

 dependencies {
+  androidTestImplementation('com.wix:detox:+')
+  implementation 'androidx.appcompat:appcompat:1.1.0'
   implementation fileTree(dir: "libs", include: ["*.jar"])
```

:::info ProGuard notice

Detox relies on
[Android Reflection API](https://developer.android.com/reference/java/lang/reflect/package-summary)
to integrate with React Native on Android, that's why you have to exempt some of its native code from
[ProGuard minification](https://developer.android.com/studio/build/shrink-code).

> Otherwise, you'll be seeing Detox crashing or hanging up infinitely upon an attempt to

Reviewer: ```suggestion
Otherwise, you'll see Detox crashing or hanging up infinitely upon an attempt to
```

run tests with your app built in **release mode**.

:::

### 3.2. Adding an auxiliary Android test

> Detox requires from your project to have a single dummy native Android test with some special content,

Reviewer: ```suggestion
Detox requires that your project has a single dummy native Android test with some special content,
```

import org.junit.runner.RunWith;

import androidx.test.ext.junit.runners.AndroidJUnit4;
import androidx.test.filters.LargeTest;
import androidx.test.rule.ActivityTestRule;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class DetoxTest {
    @Rule // (2)
// highlight-next-line
    public ActivityTestRule<MainActivity> mActivityRule = new ActivityTestRule<>(MainActivity.class, false, false);

    @Test
    public void runDetoxTests() {
        DetoxConfig detoxConfig = new DetoxConfig();
        detoxConfig.idlePolicyConfig.masterTimeoutSec = 90;
        detoxConfig.idlePolicyConfig.idleResourceTimeoutSec = 60;
        detoxConfig.rnContextLoadTimeoutSec = (BuildConfig.DEBUG ? 180 : 60);

        Detox.runTests(mActivityRule, detoxConfig);
    }
}
```

You can see there are two lines (1 and 18) that require extra care:

1. Replace with your package name. To look it up, you could copy and paste the first line from `android/app/src/main/java/com/<your.package>/MainActivity.java`.
1. Usually not the case, but you might have a custom activity name instead of a default `MainActivity`. To check whether it is so or not, open your `android/app/src/main/AndroidManifest.xml`,

> and look up how your main activity is called:

Reviewer: ```suggestion
and check what your main activity is called:
```
